### ISSUE TITLE
static route with revalidation doesn't switch between 404 and 200 status codes correctly

### ISSUE THREAD

**Issue by tdanecker** [vercel/next.js#63478]
### Link to the code that reproduces this issue

https://github.com/tdanecker/404-test

### To Reproduce

Use a static route (`/` in the linked repo) with `const export revalidate = <some number>` which, depending on a fetch request, either returns `notFound()` or the actual content.

In the linked repo:

1. Start mocked 404 API responses: `docker-compose up response-404`
2. Start app: `npm run build && npm run start`
3. Observe responses:
   first response: 404 Page, but status code 200 :x:
   subsequent responses after revalidation: 404 Page, status code 404 :white_check_mark:
4. Switch mocked API response to 200: `docker-compose up response-200`
5. Observe responses:
   first response: 404 Page, status code 404 (old stale response)
   subsequent responses: correct page content, but 404 status code :x:

### Current vs. Expected behavior

Current:
- Next.js first returns a 200 status code for a notFound / 404 page.
- The route doesn't recover correctly from a notFound / 404 page. It sends back the correct content but gets stuck at the 404 status code.

Expected:
- The first response returns the correct 404 status code for notFound.
- After the content is available, the route responds with the content and a 200 status code.

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #25~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Feb 20 16:09:15 UTC 2
  Available memory (MB): 31695
  Available CPU cores: 16
Binaries:
  Node: 20.11.0
  npm: 10.2.4
  Yarn: 3.6.1
  pnpm: 7.33.5
Relevant Packages:
  next: 14.2.0-canary.31 // Latest available version is detected (14.2.0-canary.31).
  eslint-config-next: 14.1.3
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.4.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

### Which stage(s) are affected? (Select all that apply)

next start (local), Other (Deployed)

### Additional context

On our production setup, after some time (sometimes hours), the status code seems to recover to a correct 200 status code. My theory is, that it recovers once the whole route got ejected from the full route cache.

**Comment by antimonad** [comment]
We are encountering this issue on our classifieds site currently migrating to next.js

**Comment by tdanecker** [comment]
https://github.com/vercel/next.js/pull/63731 fixes the first part of the issue. On the first request, the `notFound` response now correctly responds with a 404 status code.

**Comment by felixmusmann** [comment]
We are observing this issue as well on one of our sites, that we migrated to the app router.

**Comment by osterbaer** [comment]
Any updates here? We experience this issue on app router sites.

**Comment by DjilanMouhous** [comment]
@osterbaer Yeah I am still encountering this issue in our production environment, which is causing persistent status code errors on some paths

**Comment by mirel-v** [comment]
I'm experiencing the same issue in my project. Any advice or workarounds would be greatly appreciated!

**Comment by ARochniak** [comment]
We have resolved the stale 404 status issue by calling `revalidatePath` from the route handler. In our case, the route handler is a webhook that is triggered whenever there are changes to the content on the CMS side.

**Comment by felixmusmann** [comment]
Fully disabling the cache is not really an option for us. After some experimenting, we are now also working on a workaround involving the revalidatePath method. It irritates me though, that the on-demand revalidation seems to behave differently than the time based revalidation. Is there a reason behind this?

Nevertheless, this issue should be given a higher priority, since it might also affect the websites SEO performance. We are already seeing 404 error codes popping up in the Google Search Crawler. This is the case for our landing page as well as other pages that are main entry points for our users.

**Comment by medonat** [comment]
Is this fixed in a more recent nextjs version. Ever since we increased the automatic revalidation time of our website this problem hounts our websites?

**Comment by ackoppel** [comment]
+1

_Edit by maintainer bot: Comment was **automatically** minimized because it was considered unhelpful. (If you think this was by mistake, let us know). Please only comment if it adds context to the issue. If you want to express that you have the same problem, use the upvote 👍 on the issue description or subscribe to the issue for updates. Thanks!_

**Comment by marcoshernanz** [comment]
Thanks for the report. I re-ran the provided reproduction: the initial 404 status was fixed by #63731, and the stale status after ISR was fixed by #79512, first released in 15.4.0. The full 404 → 200 flow works correctly on 15.4.0 and current stable/canary, so I’m closing this as resolved. Please open a new issue with an updated reproduction if this still occurs on a supported version.
